### ISSUE TITLE
Custom header in browser.url('test-url')

### ISSUE THREAD

**Issue by pgangwani** [webdriverio/webdriverio#6361]
**Is your feature request related to a problem? Please describe.**
I am hard to find a way to send custom header with ```browser.url('test-url')```

**Describe the solution you'd like**
There is should be way to configure headers to be sent so that I can test canary version of code which usually tested behind header.

Please note I want the approach to work in saucelab, chromedriver & other services too.

**Comment by pgangwani** [comment]
@christian-bromann : can you triage this ?

**Comment by christian-bromann** [comment]
Any contributions to this are welcome. I suggest to have it look like this:

```js
const modifiedRequest = browser.mock('test-url')
modifiedRequest.requestWith({
  headers: { 'x-custom-headers': 'foobar' }
})

browser.url('test-url')
```

The `requestWith` command should allow to modify everything [`Fetch.continueRequest`](https://chromedevtools.github.io/devtools-protocol/tot/Fetch/#method-continueRequest) allows to do including support for [auth challenge](https://chromedevtools.github.io/devtools-protocol/tot/Fetch/#type-AuthChallengeResponse). You would need to update [this call](https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/utils/interception/devtools.ts#L174) to include this data in there.

**Comment by unickq** [comment]
what are alternatives now? tried https://webdriver.io/docs/options/#headers - no luck

**Comment by christian-bromann** [comment]
> what are alternatives now?

@unickq I laid out what is necessary to get this featured implemented. Please get involved if you have interest.

**Comment by unickq** [comment]
@christian-bromann does it mean there’s no option to set request headers now? The reason I pointed wdio docs to understand if it currently works or just planned, no offense  :)

**Comment by christian-bromann** [comment]
> does it mean there’s no option to set request headers now?

I don't understand how you can assume that this feature works if this issue literally describes that it is not.

I clarified the docs to ensure the meaning of `headers` is better explained. Again, if you are interested setting custom request headers in the browser, like this issue states, please get involved.

**Comment by ntulele** [comment]
Here are the helper methods that I use successfully, based on cdp commands:
_Caveat:_ I'm still using v5 <blush> so YMMV

```
/** WARNING!  Request interception breaks login: Apply headers post-login
 *  WARNING!  Do not use browser.reloadSession beforehand, breaks Fetch commands -- Logout instead
 *           (see webdriverio issue 5651, released in v6.3.5)
 *   WARNING!  using browser.newWindow also prevents this from working (continueRequest does not fire)
 *
 *  TIP: set logLevel: 'info' to track behavior. Actual header mod exec appears like this:
 *      devtools-service:CommandHandler: Send command "Fetch.continueRequest"
 *
 * @param urlpattern Pattern for which requests will be intercepted, and headers applied
 * @param headersArray  Example:
 *             [
 *             { name : "key",
 *               value : "value" }
 *             ]
 */
export const interceptAndApplyHeaders = (urlpattern = '*', headersArray) => {
	browser.removeAllListeners();
	try {
		browser.cdp('Fetch', 'enable', {
			patterns: [{urlPattern: urlpattern}]
		});
		console.debug("Apply headers: Fetch enabled");

		console.debug("Defining action on request interception...");
		browser.prependListener('Fetch.requestPaused', (params) => {
			browser.cdp('Fetch', 'continueRequest', {
				requestId: params.requestId,
				headers: headersArray
			});
		});
		console.info("Finished setting Fetch to intercept and apply headers");
	} catch(err) {
		console.error("error occurred during interceptAndApplyHeaders " +
				"\nActual error: \n" + err );
		if(err.toString().includes("Cannot read property 'message' of undefined")) {
			console.error("Perhaps browser.reloadSessoin used?  That breaks interceptAndApplyHeaders");
		}
		throw err;
	}
};
```


```
/** Stops request interception, and disables Fetch
 *  Apparently the continueRequest persists despite Fetch.disable, so removeAllListeners is necessary (better would be to find the right one and stop just that one)
 */
export const stopInterceptingRequests = () => {
	try {
		browser.cdp('Fetch', 'disable');
		console.info("Disabled Fetch request intercept without error");
		browser.removeAllListeners();
		console.info("Removed all listeners");
	} catch(err) {
		console.error("error occurred while attempting to disable Fetch request interception: \n" + err )
	}
};
```

**Comment by unickq** [comment]
@ntulele thanks a lot - amazing work. It really set headers everywhere, but some iframes don't use it.

Made it work with https://github.com/bewisse/modheader_selenium extension, there's a way to set it up with wdio

**Comment by ntulele** [comment]
@unickq  Thanks for pointing out that extension. How ironic that I use ModHeader Chrome extension all the time, for manual testing, but had not seen that wdio integration :)

**Comment by dichovsky** [comment]
This solution requires devtools-service, used to pass Authorization headers to each request

`export default class BasePage {
    private isInterceptorAdded: boolean;

    constructor() {
        this.isInterceptorAdded = false;
    }

    async openWithHeaders(url: string, headers: { [k: string]: string }): Promise<string> {
        if (!this.isInterceptorAdded) {
            await browser.cdp('Network', 'enable');
            await browser.cdp('Network', 'setRequestInterception', {
                patterns: [{ urlPattern: `*` }],
            });

            browser.on('Network.requestIntercepted', async (params) => {
                await browser.cdp('Network', 'continueInterceptedRequest', {
                    interceptionId: params.interceptionId,
                    headers,
                });
            });

            this.isInterceptorAdded = true;
        }

        return browser.url(url);
    }
}`

**Comment by nikaswoon** [comment]
> This solution requires devtools-service, used to pass Authorization headers to each request
> 
> ```js
> export default class BasePage {
>     private isInterceptorAdded: boolean;
> 
>     constructor() {
>         this.isInterceptorAdded = false;
>     }
> 
>     async openWithHeaders(url: string, headers: { [k: string]: string }): Promise<string> {
>         if (!this.isInterceptorAdded) {
>             await browser.cdp('Network', 'setRequestInterception', {
>                 patterns: [{ urlPattern: `*` }],
>             });
>             browser.on('Network.requestIntercepted', async (params) => {
>                 await browser.cdp('Network', 'continueInterceptedRequest', {
>                     interceptionId: params.interceptionId,
>                     headers,
>                 });
>             });
>             this.isInterceptorAdded = true;
>         }
>         return browser.url(url);
>     }
> }
> ```

Unfortunately it doesn't work for BrowserStack

**Comment by christian-bromann** [comment]
> Unfortunately it doesn't work for BrowserStack

This is barely supported in any of the cloud vendors for various reasons.

**Comment by alex-zzzz** [comment]
I found this solution to provide extra headers:
`
const puppeteer = await browser.getPuppeteer();
const pages = await puppeteer.pages();
if (pages && pages.length) {
   pages.forEach(page => {
      page.setExtraHTTPHeaders({
	'User-Agent': 'Mozilla/5.0 (iPhone; CPU iPhone OS 13_2_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.3 Mobile/15E148 Safari/604.1'
       })
   })
}
return browser.url(`url that work only with mobile headers`)`

**Comment by gorskip** [comment]
This one https://github.com/bewisse/modheader_selenium works for me.

**Comment by raktim1988** [comment]
> Any contributions to this are welcome. I suggest to have it look like this:
> 
> ```js
> const modifiedRequest = browser.mock('test-url')
> modifiedRequest.requestWith({
>   headers: { 'x-custom-headers': 'foobar' }
> })
> 
> browser.url('test-url')
> ```
> 
> The `requestWith` command should allow to modify everything [`Fetch.continueRequest`](https://chromedevtools.github.io/devtools-protocol/tot/Fetch/#method-continueRequest) allows to do including support for [auth challenge](https://chromedevtools.github.io/devtools-protocol/tot/Fetch/#type-AuthChallengeResponse). You would need to update [this call](https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/utils/interception/devtools.ts#L174) to include this data in there.

When I tried with this one, I didn't find the requestWith() function available ,Property 'requestWith' does not exist on type 'Mock'.

**Comment by christian-bromann** [comment]
@raktim1988 because this was just a suggestion how it could be implemented. This feature doesn't exist yet, please get involved if you have interest using such feature.

**Comment by gianlucamangiapelo** [comment]
Interested in this feature!

**Comment by thescripted** [comment]
> Any contributions to this are welcome. I suggest to have it look like this:
> 
> ```js
> const modifiedRequest = browser.mock('test-url')
> modifiedRequest.requestWith({
>   headers: { 'x-custom-headers': 'foobar' }
> })
> 
> browser.url('test-url')
> ```
> 
> The `requestWith` command should allow to modify everything [`Fetch.continueRequest`](https://chromedevtools.github.io/devtools-protocol/tot/Fetch/#method-continueRequest) allows to do including support for [auth challenge](https://chromedevtools.github.io/devtools-protocol/tot/Fetch/#type-AuthChallengeResponse). You would need to update [this call](https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/utils/interception/devtools.ts#L174) to include this data in there.

@christian-bromann But this approach wouldn't work in Saucelab anyway (which was the original commenter's note) because of Chrome DevTools protocol, right?

I also am interested in this feature and am actually willing to implement it if possible. Any way we can set browser headers that would also work on cloud vendors?

Off the cuff, is something like this viable:
```js
browser.url('test-url', { headers: { 'x-custom-headers': 'foobar' }})
```
or a `headers:` field within the wdio.conf.js file?

**Comment by vasiliygagin** [comment]
browser.url('test-url', { headers: { 'x-custom-headers': 'foobar' }})
would be nice

**Comment by christiansalluzzoAV** [comment]
Interested! Please and Thank you

**Comment by christian-bromann** [comment]
@christiansalluzzoAV let me set the expectation clear: if you like to have a feature, get involved! Thank you.

**Comment by niravparmar** [comment]
> @ntulele thanks a lot - amazing work. It really set headers everywhere, but some iframes don't use it.
> 
> Made it work with https://github.com/bewisse/modheader_selenium extension, there's a way to set it up with wdio

this is working, Ive tried it

**Comment by bengeorge** [comment]
Solution with https://github.com/bewisse/modheader_selenium works. Though it is critical to wait for modheader to be `Done`.

```
export const config: WebdriverIO.Config = {
  ...baseConfig,

  capabilities: [
    {
      ...
      'goog:chromeOptions': {
        extensions: [getEncodedExtension()],
      },
    },
  ],

  async before() {
    browser.url(
      getAddHeaderUrl(
        'x-vercel-protection-bypass',
        'xxxxxxxxxxxxxxxxxxxxxxxxx',
      ),
    );

    // Wait until chrome-modheader is done adding the header
    await browser.waitUntil(async () => {
      const title = await browser.getTitle();
      return title === 'Done';
    });
  },
};
```

**Comment by christian-bromann** [comment]
implemented in #13222
